reps/2024-06-16-support-apache-yunikorn-scheduler.md

```diff
@@ -0,0 +1,157 @@
+# Support new batch scheduler option: Apache YuniKorn
+
+## Motivation
+
+The [Support batch scheduling and queueing](https://github.com/ray-project/kuberay/issues/213) allows Ray to integrate
+easily with non-default Kubernetes scheduler, such as [Volcano](https://volcano.sh/). This proposal aims to add support for
+another popular Kubernetes scheduler: [Apache YuniKorn](https://yunikorn.apache.org/). This provides Ray users with 
+another option to leverage the [scheduling features](https://yunikorn.apache.org/docs/next/get_started/core_features)
+YuniKorn offers.
+
+## Existing state
+
+### Current Batch Scheduler Option
+
+Currently, the batch scheduler feature can be enabled using the `enable-batch-scheduler` boolean option.
+When set to `true`, the operator is started with the following Helm chart value override:
+
+```shell
+--set batchScheduler.enabled=true
+```
+
+The scheduler manager in the KubeRay operator will initiate the scheduler plugins. The framework provides hooks for
+each scheduler implementation to inject custom resources and modify pod metadata accordingly. When a Ray cluster is
+created, the framework calls the appropriate scheduler plugin functions based on the scheduler name provided
+by `ray.io/scheduler-name`.
+
+### Limitations
+
+The framework is designed to be scheduler-agnostic and provides general hooks for supporting different scheduler options.
+However, once the `enable-batch-scheduler` option is set to `true`, the scheduler manager will attempt to load all
+scheduler schemas by calling the `AddToScheme(scheme *runtime.Scheme)` function implemented by each scheduler plugin.
+This loads all the schemas, including CRDs defined in the implementation. Since only the `VolcanoBatchScheduler`
+is currently implemented, it always loads Volcano's `PodGroup` CRD in the controller runtime,
+requiring the installation of Volcano CRDs even when enabling other scheduler options.
+
+## Proposed changes
+
+### Targeted scheduler option
+
+To support Apache YuniKorn, and more importantly, to other schedulers in Ray.
+We propose to add an option to explicitly set the scheduler name, i.e `--batch-scheduler-name`.
+Option syntax: `--batch-scheduler-name [SupportedSchedulerName]`, for example:
+
+Option Syntax:
+
+```shell
+# 1. use volcano
+--enable-batch-scheduler=true
+--batch-scheduler-name=volcano
```

Comment:
I suggest simplifying the flags here to just a single flag:
```
--batch-scheduler=volcano|yunikorn
```

We can deprecate `--enable-batch-scheduler` for a couple releases and remove it

Reply:
Agreed